File: java/src/main/java/com/google/cloud/dataproc/templates/redshift/RedshiftToGCS.java
/*
 * Copyright (C) 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.redshift;

import static com.google.cloud.dataproc.templates.util.TemplateConstants.*;

import com.google.cloud.dataproc.templates.BaseTemplate;
import org.apache.commons.lang3.StringUtils;
import org.apache.spark.sql.DataFrameWriter;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RedshiftToGCS implements BaseTemplate {

  private static final Logger LOGGER = LoggerFactory.getLogger(RedshiftToGCS.class);

  private final String inputUrl;
  private final String inputTable;
  private final String tempDir;
  private final String iamRole;
  private final String accessKey;
  private final String secretKey;
  private final String fileFormat;
  private final String fileLocation;
  private final String outputMode;

  public RedshiftToGCS() {
    inputUrl = getProperties().getProperty(REDSHIFT_GCS_INPUT_URL);
    inputTable = getProperties().getProperty(REDSHIFT_GCS_INPUT_TABLE);
    tempDir = getProperties().getProperty(REDSHIFT_GCS_TEMP_DIR);
    iamRole = getProperties().getProperty(REDSHIFT_GCS_IAM_ROLE);
    accessKey = getProperties().getProperty(REDSHIFT_GCS_ACCESS_KEY);
    secretKey = getProperties().getProperty(REDSHIFT_GCS_SECRET_KEY);
    fileFormat = getProperties().getProperty(REDSHIFT_GCS_FILE_FORMAT);
    fileLocation = getProperties().getProperty(REDSHIFT_GCS_FILE_LOCATION);
    outputMode = getProperties().getProperty(REDSHIFT_GCS_OUTPUT_MODE);
  }

  @Override
  public void runTemplate() {

    validateInput();

    SparkSession spark = SparkSession.builder().appName("Spark HiveToGcs Job").getOrCreate();
    LOGGER.info("RedshiftToGcs job started.");

    spark.sparkContext().hadoopConfiguration().set("fs.s3a.access.key", accessKey);
    spark.sparkContext().hadoopConfiguration().set("fs.s3a.secret.key", secretKey);

    Dataset<Row> inputData =
        spark
            .read()
            .format("io.github.spark_redshift_community.spark.redshift")
            .option("url", inputUrl)
            .option("dbtable", inputTable)
            .option("tempdir", tempDir)
            .option("aws_iam_role", iamRole)
            .load();

    DataFrameWriter<Row> writer = inputData.write().format(fileFormat);
    writer.mode(outputMode).save(fileLocation);

    LOGGER.info("RedshiftToGcs job completed.");
    spark.stop();
  }

  void validateInput() {
    if (StringUtils.isAllBlank(inputUrl)
            || StringUtils.isAllBlank(inputTable)
            || StringUtils.isAllBlank(tempDir)
            || StringUtils.isAllBlank(iamRole)
            || StringUtils.isAllBlank(accessKey)
            || StringUtils.isAllBlank(secretKey)
            || StringUtils.isAllBlank(fileFormat)
            || StringUtils.isAllBlank(fileLocation)) {
      LOGGER.error(
              "{},{},{},{},{},{},{},{} is required parameter. ",
              REDSHIFT_GCS_INPUT_URL,
              REDSHIFT_GCS_INPUT_TABLE,
              REDSHIFT_GCS_TEMP_DIR,
              REDSHIFT_GCS_IAM_ROLE,
              REDSHIFT_GCS_ACCESS_KEY,
              REDSHIFT_GCS_SECRET_KEY,
              REDSHIFT_GCS_FILE_FORMAT,
              REDSHIFT_GCS_FILE_LOCATION);
      throw new IllegalArgumentException(
              "Required parameters for RedshiftToGCS not passed. "
                      + "Set mandatory parameter for RedshiftToGCS template "
                      + "in resources/conf/template.properties file.");
    }

    LOGGER.info(
            "Starting Hive to GCS spark job with following parameters:"
                    + "1. {}:{}"
                    + "2. {}:{}"
                    + "3. {}:{}"
                    + "4. {},{}"
                    + "5. {},{}"
                    + "6. {},{}"
                    + "7. {},{}"
                    + "8. {},{}",
            REDSHIFT_GCS_INPUT_URL,
            inputUrl,
            REDSHIFT_GCS_INPUT_TABLE,
            inputTable,
            REDSHIFT_GCS_TEMP_DIR,
            tempDir,
            REDSHIFT_GCS_IAM_ROLE,
            iamRole,
            REDSHIFT_GCS_ACCESS_KEY,
            accessKey,
            REDSHIFT_GCS_SECRET_KEY,
            secretKey,
            REDSHIFT_GCS_FILE_FORMAT,
            fileFormat,
            REDSHIFT_GCS_FILE_LOCATION,
            fileLocation);
  }
}

File: python/dataproc_templates/redshift/redshift_to_gcs.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, Sequence, Optional, Any
from logging import Logger
import argparse
import pprint

from pyspark.sql import SparkSession, DataFrame, DataFrameWriter

from dataproc_templates import BaseTemplate
import dataproc_templates.util.template_constants as constants

__all__ = ['RedshiftToGCSTemplate']


class RedshiftToGCSTemplate(BaseTemplate):
    """
    Dataproc template implementing loads from REDSHIFT into GCS
    """

    @staticmethod
    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
        parser: argparse.ArgumentParser = argparse.ArgumentParser()

        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_INPUT_URL}',
            dest=constants.REDSHIFTTOGCS_INPUT_URL,
            required=True,
            help='REDSHIFT input URL'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_INPUT_TABLE}',
            dest=constants.REDSHIFTTOGCS_INPUT_TABLE,
            required=True,
            help='REDSHIFT input table name'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_S3_TEMPDIR}',
            dest=constants.REDSHIFTTOGCS_S3_TEMPDIR,
            required=True,
            help='REDSHIFT S3 temporary bucket location s3a://bucket/path'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_IAM_ROLEARN}',
            dest=constants.REDSHIFTTOGCS_IAM_ROLEARN,
            required=True,
            help='REDSHIFT IAM Role with S3 Access'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_S3_ACCESSKEY}',
            dest=constants.REDSHIFTTOGCS_S3_ACCESSKEY,
            required=True,
            default="",
            help='AWS Access Keys which allow access to S3'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_S3_SECRETKEY}',
            dest=constants.REDSHIFTTOGCS_S3_SECRETKEY,
            required=True,
            default="",
            help='AWS Secret Keys which allow access to S3'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_OUTPUT_LOCATION}',
            dest=constants.REDSHIFTTOGCS_OUTPUT_LOCATION,
            required=True,
            help='GCS location for output files'
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_OUTPUT_FORMAT}',
            dest=constants.REDSHIFTTOGCS_OUTPUT_FORMAT,
            required=True,
            help='Output file format (one of: avro,parquet,csv,json)',
            choices=[
                constants.FORMAT_AVRO,
                constants.FORMAT_PRQT,
                constants.FORMAT_CSV,
                constants.FORMAT_JSON
            ]
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_OUTPUT_MODE}',
            dest=constants.REDSHIFTTOGCS_OUTPUT_MODE,
            required=False,
            default=constants.OUTPUT_MODE_APPEND,
            help=(
                'Output write mode '
                '(one of: append,overwrite,ignore,errorifexists) '
                '(Defaults to append)'
            ),
            choices=[
                constants.OUTPUT_MODE_OVERWRITE,
                constants.OUTPUT_MODE_APPEND,
                constants.OUTPUT_MODE_IGNORE,
                constants.OUTPUT_MODE_ERRORIFEXISTS
            ]
        )
        parser.add_argument(
            f'--{constants.REDSHIFTTOGCS_OUTPUT_PARTITIONCOLUMN}',
            dest=constants.REDSHIFTTOGCS_OUTPUT_PARTITIONCOLUMN,
            required=False,
            default="",
            help='GCS partition column name'
        )

        known_args: argparse.Namespace
        known_args, _ = parser.parse_known_args(args)

        return vars(known_args)

    def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:

        logger: Logger = self.get_logger(spark=spark)

        # Arguments
        input_redshift_url: str = args[constants.REDSHIFTTOGCS_INPUT_URL]
        input_redshift_tempdir: str = args[constants.REDSHIFTTOGCS_S3_TEMPDIR]
        input_redshift_table: str = args[constants.REDSHIFTTOGCS_INPUT_TABLE]
        input_redshift_iam_role: str = args[constants.REDSHIFTTOGCS_IAM_ROLEARN]
        input_redshift_accesskey: str = args[constants.REDSHIFTTOGCS_S3_ACCESSKEY]
        input_redshift_secretkey: str = args[constants.REDSHIFTTOGCS_S3_SECRETKEY]
        
        output_location: str = args[constants.REDSHIFTTOGCS_OUTPUT_LOCATION]
        output_format: str = args[constants.REDSHIFTTOGCS_OUTPUT_FORMAT]
        output_mode: str = args[constants.REDSHIFTTOGCS_OUTPUT_MODE]
        output_partitioncolumn: str = args[constants.REDSHIFTTOGCS_OUTPUT_PARTITIONCOLUMN]

        logger.info(
            "Starting REDSHIFT to GCS spark job with parameters:\n"
            f"{pprint.pformat(args)}"
        )
        
        # Read
        input_data: DataFrame
        
        spark._jsc.hadoopConfiguration().set(constants.AWS_S3ACCESSKEY, input_redshift_accesskey)
        spark._jsc.hadoopConfiguration().set(constants.AWS_S3SECRETKEY, input_redshift_secretkey)
        
        input_data=spark.read \
                .format(constants.FORMAT_REDSHIFT) \
                .option(constants.JDBC_URL, input_redshift_url) \
                .option(constants.JDBC_TABLE, input_redshift_table) \
                .option(constants.REDSHIFT_TEMPDIR, input_redshift_tempdir) \
                .option(constants.REDSHIFT_IAMROLE, input_redshift_iam_role) \
                .load()

        # Write
        if (output_partitioncolumn != ""):
            writer: DataFrameWriter = input_data.write.mode(output_mode).partitionBy(output_partitioncolumn)
        else:
            writer: DataFrameWriter = input_data.write.mode(output_mode)
            
        if output_format == constants.FORMAT_PRQT:
            writer \
                .parquet(output_location)
        elif output_format == constants.FORMAT_AVRO:
            writer \
                .format(constants.FORMAT_AVRO) \
                .save(output_location)
        elif output_format == constants.FORMAT_CSV:
            writer \
                .option(constants.HEADER, True) \
                .csv(output_location)
        elif output_format == constants.FORMAT_JSON:
            writer \
                .json(output_location)
